library/Backends/llvm/src/Juvix/Backends/LLVM/Codegen/Record.hs

```diff
@@ -1 +1,132 @@
-module Juvix.Backends.LLVM.Codegen.Record where
+module Juvix.Backends.LLVM.Codegen.Record
+  ( register,
+    loadField,
+    makeRecord,
+  )
+where
+
+import qualified Data.List as List
+import qualified Juvix.Backends.LLVM.Codegen.Block as Block
+import qualified Juvix.Backends.LLVM.Codegen.Types as Types
+import qualified Juvix.Backends.LLVM.Pass.Types as PassTypes
+import Juvix.Library hiding (Type, local)
+import qualified Juvix.Library.HashMap as Map
+import Juvix.Library.NameSymbol as NameSymbol
+import qualified LLVM.AST as AST
+import qualified LLVM.AST.Type as Type
+import qualified Prelude as P
+
+-- | @register@ registers a record type with the given name and
+-- | field names and types
+register ::
+  Types.Define m =>
+  PassTypes.RecordName ->
+  [PassTypes.FieldName] ->
+  [Type.Type] ->
+  m ()
+register recordName fieldNames llvmFieldTypes = do
+  let llvmTypeName = recordTypeName recordName
+  Block.addType llvmTypeName (llvmRecordType llvmFieldTypes)
+  recordTable <- get @"recordTab"
+  let typeRef = Type.NamedTypeReference llvmTypeName
+  let fieldDescs = zip (map toSymbol fieldNames) llvmFieldTypes
+  put @"recordTab" $ Map.insert (toSymbol recordName) (typeRef, fieldDescs) recordTable
+
+recordTypeName :: NameSymbol.T -> AST.Name
+recordTypeName recordName = Block.internName $ "record-" <> toSymbol recordName
+
+llvmRecordType :: [Type.Type] -> Type.Type
+llvmRecordType llvmFieldTypes =
+  Type.StructureType
+    { isPacked = True,
+      elementTypes =
+        -- We box types, both to allow recursive structures and also to avoid
+        -- having to perform any recursive size calculations for now.
+        map Types.pointerOf llvmFieldTypes
+    }
+
+-- | Get the names and types of the fields of the record type with the given name.
+recordFields :: Types.Define m => PassTypes.RecordName -> m [(Symbol, Type.Type)]
+recordFields recordName = do
+  recordTable <- get @"recordTab"
+  pure $
+    snd $
+      fromMaybe (P.error "typechecker allowed record of non-existent type") $
+        Map.lookup (toSymbol recordName) recordTable
+
+-- | Get the index of the field with the given name within the list of
+-- | fields of the record with the given name.
+fieldIndex ::
+  Types.Define m =>
+  PassTypes.RecordName ->
+  PassTypes.FieldName ->
+  m Int
+fieldIndex recordName fieldName = do
+  fieldDescs <- recordFields recordName
+  pure $
+    fromMaybe (P.error "typechecker allowed selection of non-existent field") $
+      List.elemIndex (toSymbol fieldName) $ map fst fieldDescs
```

Comment:
Same here, as long as we are in the monad, we can exhibit errors for no cost properly.

Reply:
Also done in https://github.com/anoma/juvix/pull/1106/commits/fc0975b6c7634a2e6496686201045cfe56d7e113 and https://github.com/anoma/juvix/pull/1106/commits/9385ae17c907812daaa88164ad3a2e17fab2f42b.